File: bindings/python/hako_binary/binary_reader.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import json
import sys
import base64

from hako_binary import binary_io
from hako_binary import offset_parser

def decode_base64(data):
    return base64.b64decode(data)

def binary_read(offmap, typename, binary_data):
    json_data = {}
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        meta = binary_io.PduMetaData()
        meta.set_empty()
        #print("binary_data: size", len(binary_data))
        #print("meta.to_bytes(): size", len(meta.to_bytes()))
        binary_io.writeBinary(binary_data, 0, meta.to_bytes())
    binary_read_recursive(meta, offmap, binary_data, json_data, binary_io.PduMetaData.PDU_META_DATA_SIZE, typename)
    return json_data

def binary_read_recursive(meta: binary_io.PduMetaData, offmap, binary_data, json_data, base_off, typename):
    #lines = offmap[typename]
    lines = offmap.get(typename)
    for line in lines:
        off = offset_parser.member_off(line) + base_off
        type = offset_parser.member_type(line)
        name = offset_parser.member_name(line)
        size = offset_parser.member_size(line)
        if (offset_parser.is_primitive(line)):
            if (offset_parser.is_single(line)):
                bin = binary_io.readBinary(binary_data, off, size)
                value = binary_io.binTovalue(type, bin)
                json_data[name] = value
            elif (offset_parser.is_array(line)):
                array_value = binary_io.readBinary(binary_data, off, size)
                json_data[name + '__raw' ] = array_value
                json_data[name] = binary_io.binToArrayValues(type, array_value)
            else: #varray
                array_size = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, off, 4))
                offset_from_heap = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, off + 4, 4))
                one_elm_size = size
                array_value = binary_io.readBinary(binary_data, meta.heap_off + offset_from_heap, one_elm_size * array_size)
                json_data[name + '__raw' ] = array_value
                json_data[name] = binary_io.binToArrayValues(type, array_value)
        else:
            if (offset_parser.is_single(line)):
                tmp_json_data = {}
                binary_read_recursive(meta, offmap, binary_data, tmp_json_data, off, type)
                json_data[name] = tmp_json_data
            elif (offset_parser.is_array(line)):
                i = 0
                array_size = offset_parser.array_size(line)
                one_elm_size = int(size / array_size)
                array_value = []
                while i < array_size:
                    tmp_json_data = {}
                    binary_read_recursive(meta, offmap, binary_data, tmp_json_data, off + (i * one_elm_size), type)
                    array_value.append(tmp_json_data)
                    i = i + 1
                json_data[name] = array_value
            else: #varray
                array_size = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, off, 4))
                offset_from_heap = binary_io.binTovalue("int32", binary_io.readBinary(binary_data, off + 4, 4))
                one_elm_size = size
                i = 0
                array_value = []
                while i < array_size:
                    tmp_json_data = {}
                    binary_read_recursive(meta, offmap, binary_data, tmp_json_data, meta.heap_off + offset_from_heap + (i * one_elm_size), type)
                    array_value.append(tmp_json_data)
                    i = i + 1
                json_data[name] = array_value

File: bindings/elixir_project/lib/hako_binary/binary_reader.ex
defmodule HakoBinary do
  alias PduMetaData
  alias OffsetParser

  def decode_base64(data) do
    :base64.decode(data)
  end

  def binary_read(offmap, typename, binary_data) do
    json_data = %{}
    meta = PduMetaDataParser.load_pdu_meta(binary_data)

    meta =
      case meta do
        nil ->
          meta = PduMetaData.new()
          PduMetaData.set_empty(meta)
          PduConversion.write_binary(binary_data, 0, PduMetaData.to_bytes(meta))
          meta

        _ -> meta
      end

    binary_read_recursive(meta, offmap, binary_data, json_data, PduMetaData.PDU_META_DATA_SIZE, typename)
    json_data
  end

  defp binary_read_recursive(meta, offmap, binary_data, json_data, base_off, typename) do
    lines = OffsetMap.get(offmap, typename)

    for line <- lines do
      off = OffsetParser.member_off(line) + base_off
      type = OffsetParser.member_type(line)
      name = OffsetParser.member_name(line)
      size = OffsetParser.member_size(line)

      cond do
        OffsetParser.is_primitive(line) ->
          cond do
            OffsetParser.is_single(line) ->
              bin = PduConversion.read_binary(binary_data, off, size)
              value = PduConversion.bin_to_value(type, bin)
              Map.put(json_data, name, value)

            OffsetParser.is_array(line) ->
              array_value = PduConversion.read_binary(binary_data, off, size)
              json_data = Map.put(json_data, "#{name}__raw", array_value)
              Map.put(json_data, name, PduConversion.bin_to_array_values(type, array_value))

            true ->  # varray
              array_size = PduConversion.bin_to_value("int32", PduConversion.read_binary(binary_data, off, 4))
              offset_from_heap = PduConversion.bin_to_value("int32", PduConversion.read_binary(binary_data, off + 4, 4))
              one_elm_size = size
              array_value = PduConversion.read_binary(binary_data, meta.heap_off + offset_from_heap, one_elm_size * array_size)
              json_data = Map.put(json_data, "#{name}__raw", array_value)
              Map.put(json_data, name, PduConversion.bin_to_array_values(type, array_value))
          end

        true -> # nested or array of structs
          cond do
            OffsetParser.is_single(line) ->
              tmp_json_data = %{}
              binary_read_recursive(meta, offmap, binary_data, tmp_json_data, off, type)
              Map.put(json_data, name, tmp_json_data)

            OffsetParser.is_array(line) ->
              array_size = OffsetParser.array_size(line)
              one_elm_size = div(size, array_size)
              array_value = for i <- 0..(array_size - 1) do
                tmp_json_data = %{}
                binary_read_recursive(meta, offmap, binary_data, tmp_json_data, off + (i * one_elm_size), type)
                tmp_json_data
              end
              Map.put(json_data, name, array_value)

            true ->  # varray
              array_size = PduConversion.bin_to_value("int32", PduConversion.read_binary(binary_data, off, 4))
              offset_from_heap = PduConversion.bin_to_value("int32", PduConversion.read_binary(binary_data, off + 4, 4))
              one_elm_size = size
              array_value = for i <- 0..(array_size - 1) do
                tmp_json_data = %{}
                binary_read_recursive(meta, offmap, binary_data, tmp_json_data, meta.heap_off + offset_from_heap + (i * one_elm_size), type)
                tmp_json_data
              end
              Map.put(json_data, name, array_value)
          end
      end
    end
    json_data
  end
end
